# 03 — Эксперименты

Сравнение 5 моделей на едином стратифицированном сплите (seed из `.env`).

In [1]:
import matplotlib.pyplot as plt
import pandas as pd

from src.config import settings
from src.modeling import run_experiments, train_and_eval_baseline

In [2]:
baseline = train_and_eval_baseline()
experiments = run_experiments()
reports = {**baseline, **experiments}
table = pd.DataFrame({k: v.to_dict() for k, v in reports.items()}).T
table = table.sort_values('roc_auc', ascending=False)
table

/Users/andrejsvistunov/Desktop/HSE/ML/hseml-group-project-andreysvistunovhsemiem-2/.venv/lib/python3.13/site-packages/sklearn/utils/validation.py:2691: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
/Users/andrejsvistunov/Desktop/HSE/ML/hseml-group-project-andreysvistunovhsemiem-2/.venv/lib/python3.13/site-packages/sklearn/utils/validation.py:2691: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


,roc_auc,pr_auc,f1,precision,recall,accuracy,tn,fp,fn,tp
xgboost,0.774466,0.551359,0.468127,0.659808,0.362745,0.817621,4419.0,248.0,845.0,481.0
logreg_fe,0.768128,0.536724,0.519896,0.458046,0.601056,0.754380,3724.0,943.0,529.0,797.0
lightgbm,0.766012,0.541377,0.517832,0.491204,0.547511,0.774403,3915.0,752.0,600.0,726.0
random_forest,0.763560,0.530337,0.448900,0.638387,0.346154,0.811947,4407.0,260.0,867.0,459.0
logreg_baseline,0.762882,0.536221,0.528541,0.496032,0.565611,0.776740,3905.0,762.0,576.0,750.0
knn_k25,0.749040,0.506966,0.435910,0.643594,0.329563,0.811280,4425.0,242.0,889.0,437.0
dummy_stratified,0.500399,0.221396,0.221045,0.221884,0.220211,0.656599,3643.0,1024.0,1034.0,292.0


In [ ]:
settings.images_dir.mkdir(parents=True, exist_ok=True)
fig, ax = plt.subplots(figsize=(7, 3.5))
table['roc_auc'].plot(kind='barh', ax=ax, color='#4a7ab5')
ax.set_xlabel('ROC-AUC')
ax.set_title(f'ROC-AUC по моделям (test, seed={settings.random_state})')
ax.invert_yaxis()
fig.tight_layout(); fig.savefig(settings.images_dir / 'experiments_roc_auc.png', dpi=120); plt.show()

## Выводы

- Baseline LogReg без FE: ROC-AUC ≈ 0.763.
- FE даёт линейной модели +0.005 ROC-AUC (`logreg_fe`).
- KNN (k=25) сопоставим с LogReg по ROC-AUC, но сильно теряет в recall.
- Ансамбли деревьев — сильнее линейных: RF ≈ 0.764, LightGBM ≈ 0.766, XGBoost ≈ 0.774 (лучший результат).
- Выбор финальной модели + тюнинг — в CP2.